engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java

```diff
@@ -125,6 +126,21 @@ public static Vector fromArray(Object arr) {
     return new Vector(arr);
   }
 
+  @TruffleBoundary
+  public Array toEnsoArray() throws InvalidArrayIndexException {
+    var interop = InteropLibrary.getUncached();
+    try {
+      long size = interop.getArraySize(storage);
+      Object[] arr = new Object[Math.toIntExact(size)];
+      for (int i = 0; i < size; i++) {
+        arr[i] = interop.readArrayElement(storage, i);
+      }
+      return new Array(arr);
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException("Unreachable", e);
+    }
+  }
```

Comment:
Is copying the array _on each invocation_ really the right solution? That seems quite costly.

I know that currently its only in `File` where the arrays will be small so it should probably be fine.

But I'm worried that even _having_ this logic will encourage using it in some other places where the performance price can be significant.

Shouldn't we instead remove the `expectArray` altogether from our TypesGen and actually use interop (`readArrayElement`) to access the elements? Like we do e.g. in `sort`?

Reply:
Can we not use the `CoerceArrayNode` which will covert as needed and avoids copying?

Its what we use in the other places like `Polyglot.execute`?

Reply:
This method is called only from `*MethodGen` classes, generated for builtin methods by the `MethodProcessor`, and only in case when then builtin method, on the Java side, takes a plain Java array as the argument, as in [EnsoFile.java:inputStream](https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java#L57C26-L57C26), and the actual argument is not of type `Array`. Which makes it rather rare.

I understand your hesitance, I don't like this solution as well, but it is a compromise because tweaking builtin method processor such that it does not copy stuff from `Vector` into `Array` would require much more work.

As backend devs, we are responsible for builtin methods, and we should know that for every builtin method, a `RootNode` is generated along with one more `Node`, meaning that we can take full advantage of Truffle stuff. If you want to accept `OpenOption[]` arguments from a polyglot array, you still need to copy it and convert it to a host object. If we intend to optimize that, we should optimize that in the builtin method - by taking `Object` and treating it as a polyglot array - rather than trying to solve this in general in builtin method processor.

Reply:
> but it is a compromise because tweaking builtin method processor such that it does not copy stuff from `Vector` into `Array` would require much more work.

I did not want to suggest that.

Rather, I'd not allow _any_ 'automatic' solution and require the creator of a particular builtin to handle this. IMO this is more explicit and then makes it clearer how each node is handling this (copying or not).

With an 'automatic' solution like this one, the copying is more hidden and I'm worried someone may forget about this when adding some builtins (especially as the team grows, such implicit knowledge can easily not get passed forward).

Reply:
> I understand your hesitance, I don't like this solution as well, but it is a compromise because tweaking builtin method processor such that it does not copy stuff from Vector into Array would require much more work.

I don't like the solution either, but it fixes the problem.

Reply:
I was imagining we would do something more like:

```java
  @Builtin.Method(name = "move_builtin", description = "Move this file to a target destination")
  @Builtin.WrapException(from = IOException.class)
  @Builtin.WrapException(from = IllegalArgumentException.class)
  @CompilerDirectives.TruffleBoundary
  public void move(EnsoFile target, Object options, @CachedLibrary("this") CoerceArrayNode coerceNode) throws IOException {
    var objectArray = coerceNode.execute(options);
    var parsedOptions = new CopyOption[objectArray.length];
    for (int i = 0; i < objectArray.length; i++) {
      if (!(objectArray[i] instanceof CopyOption)) {
        throw new IllegalArgumentException("options should be a Vector CopyOptions.");
      }

      parsedOptions[i] = (CopyOption) objectArray[i];
    }

    truffleFile.move(target.truffleFile, parsedOptions);
  }
```

Additionally, I thought publishing Java types from the built-ins was not recommended?

Reply:
+1 to @jdunkerley suggestion - I also thought we could have done this way and it would be cleaner and safer then. Are there any disadvantages to such solution?

Reply:
I have reverted this controversial modification at the end, and rather implemented something similar to your suggestion. The builtin method processor now fails with an error when processing a builtin method with a parameter of an array of non-primitive, non-guest types. I have updated the description of the PR.